File: Python/Visitor/Gemini 1.5 Flash/round2/code.py
from abc import ABC, abstractmethod
from typing import List
import datetime

class ATM:
    def __init__(self):
        self.users = {}  # {card_number: {"pin": pin, "balance": balance}}
        self.max_pin_attempts = 3

    def add_user(self, card_number, pin, balance):
        self.users[card_number] = {"pin": pin, "balance": balance}

    def authenticate(self, card_number, pin, attempts=0):
        user = self.users.get(card_number)
        if user and user["pin"] == pin:
            return True
        elif attempts >= self.max_pin_attempts:
            print("Card retained due to too many incorrect PIN attempts.")
            return False
        else:
            print("Incorrect PIN. Please try again.")
            return self.authenticate(card_number, pin, attempts + 1)

    def process_transaction(self, card_number, visitor):
        if self.authenticate(card_number, visitor.get_pin()):
            user = self.users[card_number]
            visitor.execute(user)
            if visitor.should_print_receipt():
                self.print_receipt(user, visitor.get_transaction_details())

    def print_receipt(self, user, transaction_details):
        print("-" * 20)
        print(f"Transaction Details: {transaction_details}")
        print(f"Date: {datetime.datetime.now().strftime('%Y-%m-%d %H:%M:%S')}")
        print(f"Balance: {user['balance']}")
        print("-" * 20)

class ATMVisitor(ABC):
    @abstractmethod
    def execute(self, user):
        pass

    @abstractmethod
    def get_pin(self):
        pass

    @abstractmethod
    def should_print_receipt(self):
        pass

    @abstractmethod
    def get_transaction_details(self):
        pass

class BalanceInquiryVisitor(ATMVisitor):
    def __init__(self, pin):
        self.pin = pin
        self.receipt_flag = False
        self.transaction_details = None

    def execute(self, user):
        print(f"Your current balance is: {user['balance']}")
        self.transaction_details = f"Balance Inquiry: {user['balance']}"

    def get_pin(self):
        return self.pin

    def should_print_receipt(self):
        return self.receipt_flag

    def get_transaction_details(self):
        return self.transaction_details

class CashWithdrawalVisitor(ATMVisitor):
    def __init__(self, pin, amount):
        self.pin = pin
        self.amount = amount
        self.receipt_flag = True
        self.transaction_details = None

    def execute(self, user):
        if user['balance'] >= self.amount:
            user['balance'] -= self.amount
            print(f"Cash withdrawn: {self.amount}")
            self.transaction_details = f"Cash Withdrawal: {self.amount}"
        else:
            print("Insufficient funds.")

    def get_pin(self):
        return self.pin

    def should_print_receipt(self):
        return self.receipt_flag

    def get_transaction_details(self):
        return self.transaction_details

class DepositVisitor(ATMVisitor):
    def __init__(self, pin, amount):
        self.pin = pin
        self.amount = amount
        self.receipt_flag = True
        self.transaction_details = None

    def execute(self, user):
        user['balance'] += self.amount
        print(f"Amount deposited: {self.amount}")
        self.transaction_details = f"Deposit: {self.amount}"

    def get_pin(self):
        return self.pin

    def should_print_receipt(self):
        return self.receipt_flag

    def get_transaction_details(self):
        return self.transaction_details

class PINChangeVisitor(ATMVisitor):
    def __init__(self, old_pin, new_pin):
        self.old_pin = old_pin
        self.new_pin = new_pin
        self.receipt_flag = True
        self.transaction_details = None

    def execute(self, user):
        if user['pin'] == self.old_pin:
            if self.new_pin == self.confirm_new_pin():
                user['pin'] = self.new_pin
                print("PIN changed successfully.")
                self.transaction_details = "PIN Change"
            else:
                print("New PIN mismatch.")
        else:
            print("Incorrect old PIN.")

    def get_pin(self):
        return self.old_pin

    def should_print_receipt(self):
        return self.receipt_flag

    def get_transaction_details(self):
        return self.transaction_details

    def confirm_new_pin(self):
        return input("Confirm new PIN: ")

File: Java/Visitor/Gemini 1.5 Flash/round2/code.java
// ATM.java
public class ATM {
    private Account account;
    private ATMVisitor visitor;

    public ATM(Account account) {
        this.account = account;
        this.visitor = new ATMVisitor(account);
    }

    public boolean authenticate(Card card, String pin) {
        if (card.getAccount().equals(account) && account.verifyPIN(pin)) {
            return true;
        }
        return false;
    }

    public void performTransaction(Transaction transaction) {
        transaction.accept(visitor);
    }

    public void printReceipt(Transaction transaction) {
        System.out.println("Receipt:");
        System.out.println("Date: " + transaction.getDate());
        System.out.println("Time: " + transaction.getTime());
        System.out.println("Transaction: " + transaction.getDescription());
        System.out.println("Amount: " + transaction.getAmount());
        System.out.println("Balance: " + account.getBalance());
    }
}

// Account.java
public class Account {
    private double balance;
    private String pin;

    public Account(double balance, String pin) {
        this.balance = balance;
        this.pin = pin;
    }

    public double getBalance() {
        return balance;
    }

    public void setBalance(double balance) {
        this.balance = balance;
    }

    public String getPIN() {
        return pin;
    }

    public void setPIN(String pin) {
        this.pin = pin;
    }

    public boolean verifyPIN(String enteredPIN) {
        return enteredPIN.equals(pin);
    }
}

// Card.java
public class Card {
    private String cardNumber;
    private Account account;

    public Card(String cardNumber, Account account) {
        this.cardNumber = cardNumber;
        this.account = account;
    }

    public Account getAccount() {
        return account;
    }
}

// Transaction.java
public abstract class Transaction {
    private String description;
    private double amount;
    private String date;
    private String time;

    public Transaction(String description, double amount) {
        this.description = description;
        this.amount = amount;
        this.date = new SimpleDateFormat("yyyy-MM-dd").format(new Date());
        this.time = new SimpleDateFormat("HH:mm:ss").format(new Date());
    }

    public String getDescription() {
        return description;
    }

    public double getAmount() {
        return amount;
    }

    public String getDate() {
        return date;
    }

    public String getTime() {
        return time;
    }

    public abstract void accept(Visitor visitor);
}

// BalanceDisplay.java
public class BalanceDisplay extends Transaction {
    public BalanceDisplay() {
        super("Balance Display", 0);
    }

    @Override
    public void accept(Visitor visitor) {
        visitor.visit(this);
    }
}

// CashWithdrawal.java
public class CashWithdrawal extends Transaction {
    public CashWithdrawal(double amount) {
        super("Cash Withdrawal", amount);
    }

    @Override
    public void accept(Visitor visitor) {
        visitor.visit(this);
    }
}

// Deposit.java
public class Deposit extends Transaction {
    public Deposit(double amount) {
        super("Deposit", amount);
    }

    @Override
    public void accept(Visitor visitor) {
        visitor.visit(this);
    }
}

// PINChange.java
public class PINChange extends Transaction {
    private String newPIN;

    public PINChange(String newPIN) {
        super("PIN Change", 0);
        this.newPIN = newPIN;
    }

    public String getNewPIN() {
        return newPIN;
    }

    @Override
    public void accept(Visitor visitor) {
        visitor.visit(this);
    }
}

// Visitor.java
public interface Visitor {
    void visit(BalanceDisplay transaction);
    void visit(CashWithdrawal transaction);
    void visit(Deposit transaction);
    void visit(PINChange transaction);
}

// ATMVisitor.java
public class ATMVisitor implements Visitor {
    private Account account;

    public ATMVisitor(Account account) {
        this.account = account;
    }

    @Override
    public void visit(BalanceDisplay transaction) {
        System.out.println("Your balance is: " + account.getBalance());
    }

    @Override
    public void visit(CashWithdrawal transaction) {
        if (account.getBalance() >= transaction.getAmount()) {
            account.setBalance(account.getBalance() - transaction.getAmount());
            System.out.println("Cash withdrawn successfully.");
        } else {
            System.out.println("Insufficient funds.");
        }
    }

    @Override
    public void visit(Deposit transaction) {
        account.setBalance(account.getBalance() + transaction.getAmount());
        System.out.println("Deposit successful.");
    }

    @Override
    public void visit(PINChange transaction) {
        // Implement PIN change logic with confirmation
        account.setPIN(transaction.getNewPIN());
        System.out.println("PIN changed successfully.");
    }
}